### PR TITLE
always close memmaps with the asdf file

### PR DESCRIPTION
**Pull request by braingram**
even when reading an already opened file

fixes issue #1239

update changelog

add regression test docstring

### CODE REVIEW

File: asdf/generic_io.py [hunk 1]
```diff
@@ -779,11 +779,9 @@ def read_into_array(self, size):
         return np.fromfile(self._fd, dtype=np.uint8, count=size)
 
     def close(self):
+        self.flush_memmap()
         super().close()
-        if self._close:
-            if hasattr(self, "_mmap") and not self._mmap.closed:
-                self._mmap.flush()
-                self._mmap.close()
```
Comment (WilliamJamieson) [inline asdf/generic_io.py:786]: Are there any additional implications of this change?

Comment (braingram) [reply asdf/generic_io.py:786]: None that I've been able to uncover. It does not appear that any other tests aside from the one added in this PR and the ones in astropy (that load a table from asdf) suffer from this issue. With 2.13 memmaps were handled by numpy.memmap which has no api for closing the memmap (as per the docstring here: https://github.com/numpy/numpy/blob/v1.23.0/numpy/core/memmap.py#L40-L43) so it's difficult if not impossible to say we've mimicked the old behavior.
Is there something specific that you had in mind?

Comment (WilliamJamieson) [reply asdf/generic_io.py:786]: No, it seems like a reasonable change then.

**Comment by braingram** [comment]
The added regression test fails on with asdf 2.14 (and the current master branch) with both pytest-openfiles 0.4.0 and 0.5.0 (as it checks for the number of open files directly instead of relying on the plugin). The test passes with this PR.

I'm inclined to ignore the minor coverage change (-0.01%) but am happy to attempt to update the PR with more coverage if that's preferred.

**Review by WilliamJamieson — APPROVED** [review]

**Comment by WilliamJamieson** [comment]
> I'm inclined to ignore the minor coverage change (-0.01%) but am happy to attempt to update the PR with more coverage if that's preferred.

Coverage seems to fluctuate for strange reasons at times. I didn't see any lines drop coverage so its no big deal.

**Pull request merged**